Extract Design Grid to design-picker component

ciampo opened the pull request:
#### Changes proposed in this Pull Request

* Extract the `<DesignPicker />` component to a new `@automattic/design-picker` package.
* Integrate `<DesignPicker />` and all design and font related utils with Gutenboarding.
* Update the script that generates the design screenshots for Gutenboarding

#### Technical updates

1. Add `DesignPicker` component and its dependencies to the new package.
  * add available-designs-config.json file
  * extract `DesignPicker` styles from `DesignSelector` in Gutenboarding. 

2. `@automattic/design-picker` package exposes also design and font related constants, types and utils as the public API.

3. Integrate `DesignPicker` package with Gutenboarding.
  * Cleanup constants.ts and stores/onboard/types
  * Update imports of all design and font related types and constants
  * Import getAvailableDesigns fn and call it instead of importing availableDesigns
  * Update `DesignSelector` step in Gutenboarding to render DesignPicker and pass randomized designs, premium badge, and other needed props to it.
  * Add placeholder integration and unit tests

#### Testing instructions

* `/new/design` should look and work the same as in production
* sites created with `/new` flow should have the correct design

* Make sure that the script that generates the design screenshots for Gutenboarding works as expected

Inspired from spike #51226

Fixes #51229

ciampo committed:
Create Design Picker component

ciampo committed:
Add design picker to the list of `client` dependencies

Commit:
Add DesignPicker component and its dependencies to the package.

* update tests and mocks including folder names
* update package.json and tsconfig.json
* add available-designs-config.json file
* extract DesignPicker styles from DesignSelector in Gutenboarding
* export design and font related constants as the public API of the package

Commit:
Integrate DesignPicker package in Gutenboarding.

* Cleanup constants.ts and stores/onboard/types
* Update imports of all design and font related types and constants
* Import getAvailableDesigns fn and call it instead of importing availableDesigns
* Update DesignSelector step to render DesignPicker and pass randomized designs, premium badge and other needed props to it.

Commit:
fix DesignPicker integration test

razvanpapadopol commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -64,84 +57,30 @@ const DesignSelector: React.FunctionComponent = () => {
+						<JetpackLogo className="design-selector__premium-badge-logo" size={ 20 } />
+						<span className="design-selector__premium-badge-text">{ __( 'Premium' ) }</span>
+					</Badge>
+				}
```
This badge is not used in production and it can be displayed only by manually switching `is_premium` property to `true` in the `available-designs-config.json` for a design. Here's a screen capture with the badge:
<img width="310" alt="Screenshot 2021-03-24 at 15 02 07" src="https://user-images.githubusercontent.com/14192054/112328710-43426900-8cbf-11eb-9c58-a89c91d3b5cd.png">

ciampo committed:
Simplify styles imports

~@wordpress/* imports are not needed since they are already imported
by the ~@automattic/onboarding/* imports

StefanNieuwenhuis commented:
### Notes


- [x] include styles from @ciampo PR --> `packages/design-picker/src/components/style.scss`.
- [x] investigate relative paths in `tsconfig.json` (`i18n-utils` in particular).
- [x] - update dependencies in `package.json`
- [x] export types to `design-picker/src/types.ts`
- [x] review available-design file
- [x] `bin/generate-gutenboarding-design-thumbnails.js`.
- [x] Investigate utils exports (related to the bin script) in `design-picker/src/index.ts`.

ciampo committed:
Clean and sort @automattic/* dependencies in package.json and tsconfig.json

ciampo committed:
Update dependencies versions, move react to peerDeps

ciampo committed:
Move mshots tests under __tests__/*

ciampo committed:
Split types, constants and utils across different files

ciampo committed:
Rename constants using SNAKE_UPPERCASE naming convention

ciampo committed:
Export pre-computed "availableDesigns" and "availableDesignsConfig"

ciampo committed:
Extract DESIGN_IMAGE_FOLDER and export from the package

ciampo committed:
Attempt to refactor the generate-gutenboarding-design-thumbnails.js script

ciampo commented:
Hey @Automattic/ganon, as part of this PR we would need to make some changes to the `bin/generate-gutenboarding-design-thumbnails.js` script.

Can you review our changes and let us know if there's anything else to do?

Also, the script on my machine is not working, but the error seems related to the `sharp` package (and not to the changes in this PR):

```
(node:4975) UnhandledPromiseRejectionWarning: Error: Invalid input
    at Sharp (/Users/marcociampini/Code/wp-calypso/node_modules/sharp/lib/constructor.js:157:11)
    at /Users/marcociampini/Code/wp-calypso/bin/generate-gutenboarding-design-thumbnails.js:102:23
    at Array.forEach (<anonymous>)
    at /Users/marcociampini/Code/wp-calypso/bin/generate-gutenboarding-design-thumbnails.js:101:22
    at async Promise.all (index 0)
    at async run (/Users/marcociampini/Code/wp-calypso/bin/generate-gutenboarding-design-thumbnails.js:65:2)
```

ciampo commented on bin/install-gutenboarding-design-thumbnails-dependencies.js:
```diff
@@ -46,3 +46,9 @@ if ( installDependencies.status ) {
 const touchDate = new Date();
 fs.utimesSync( 'node_modules', touchDate, touchDate );
+
+// @TODO: error handling
+execSync( 'cd packages/design-picker && yarn build', {
+	shell: true,
+	stdio: 'inherit',
+} );
```
Make sure that `@automattic/design-picker` is installed

ciampo commented on bin/generate-gutenboarding-design-thumbnails.mjs:
```diff
@@ -12,19 +12,24 @@
+import captureWebsite from 'capture-website';
+import sharp from 'sharp';
+import wpUrl from '@wordpress/url';
+import { availableDesignsConfig, DESIGN_IMAGE_FOLDER } from '@automattic/design-picker';
```
Since the `available-designs-config.json` file has been moved to the `@automattic/design-picker` package, we need to update the imports.

But it seems that Calypso packages don't expose a `cjs` build, and so:
  - we updated this file to use ES modules syntax, and changed the file extension to `.mjs`
  - we [added a step in `bin/install-gutenboarding-design-thumbnails-dependencies.js`](https://github.com/Automattic/wp-calypso/pull/51306/files#r601641351) to make sure that the `@automattic/design-picker` component is built

At the moment, when running `node ./bin/generate-gutenboarding-design-thumbnails.mjs` on my machine, I get the following error:

```
/Users/marcociampini/Code/wp-calypso/node_modules/@wordpress/components/build/scroll-lock/index.js:41
      htmlDocument = _ref$htmlDocument === void 0 ? document : _ref$htmlDocument,
                                                    ^

ReferenceError: document is not defined
    at createScrollLockComponent (/Users/marcociampini/Code/wp-calypso/node_modules/@wordpress/components/build/scroll-lock/index.js:41:53)
    at Object.<anonymous> (/Users/marcociampini/Code/wp-calypso/node_modules/@wordpress/components/build/scroll-lock/index.js:153:16)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/Users/marcociampini/Code/wp-calypso/node_modules/@wordpress/components/build/popover/index.js:36:42)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
```

ciampo replied:
Update: I removed these changes in favour of a simpler (but potentially more hacky) solution across these commits:
- https://github.com/Automattic/wp-calypso/pull/51306/commits/d53f83e4a64a2de1bbe92abdd613f5c5424280fd
- https://github.com/Automattic/wp-calypso/pull/51306/commits/66c4677e0c3c52374e3d1b51bfd316eb5954bb3d
- https://github.com/Automattic/wp-calypso/pull/51306/commits/4d8da2a80444c5f8f263e52a1dcb6d1761fc6d36

p-jackson replied:
> But it seems that Calypso packages don't expose a `cjs` build

That's interesting, they have `tsconfig-cjs.json` that should be producing a `cjs` build. Maybe that doesn't happen automatically and is only there just in case we want to publish to npm.

I think using a relative path to `available-designs-config.json` is fine since it's in a monorepo and this is just a script that's always supposed to be run from within the monorepo.

StefanNieuwenhuis replied:
Thanks for the heads-up. When the script only runs within the monorepo, a relative path should be fine!

razvanpapadopol commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -139,7 +139,6 @@ function waitForSelectedSite(): Promise< Site | undefined > {
  */
 function ensureRandomizedDesignsAreUpToDate() {
 	const designsInStore = select( ONBOARD_STORE ).getRandomizedDesigns();
-	const availableDesigns = getAvailableDesigns();
```
@ciampo, could you provide some context for rolling back this change? When we'll be implementing the first A/C of https://github.com/Automattic/wp-calypso/issues/51404 we'll need a way to call `getAvailableDesigns` with some argument so it would return the randomized list of designs. Of course, exposing both the function and the pre-computed list of designs could be an option.

ciampo replied:
We rolled it back because we're exposing `availableDesigns` directly (see line 14), which is basically the same, but with the computation done only once directly in the `@automattic/design-picker` module.

It felt like a cleaner solution, especially as we were not aware of the A/Cs of future issues like  #51404. I can roll this change back if you prefer.

razvanpapadopol replied:
It's fine to leave it like this. Ultimately, I'm 100% for keeping the scope of this extraction PR somewhat contained. We'll cross that bridge when we come to it.

razvanpapadopol commented on packages/design-picker/src/utils/available-designs.ts:
```diff
@@ -57,31 +56,18 @@ export const mShotOptions = (): MShotsOptions => {
 const canUseWebP = getCanUseWebP();
 
 export const getDesignImageUrl = ( design: Design ): string => {
```
I know it's not really the scope of this PR but since we're extracting this function to a package, let's try to make more clear what should be updated in order for the cache busting to happen.

```suggestion
// Bump the version query param here to cache bust the images after running bin/generate-gutenboarding-design-thumbnails.js
export const getDesignImageUrl = ( design: Design ): string => {
```

ciampo committed:
Revert changes to the the generate-gutenboarding-design-thumbnails.js script

ciampo committed:
Attempt a more hacky refactor for the generate-gutenboarding-design-thumbnails.js script

ciampo committed:
Rename script file back to .js extension

ciampo committed:
Update packages/design-picker/src/utils/available-designs.ts

Co-authored-by: Răzvan Papadopol <razvan.papadopol@automattic.com>

razvanpapadopol commented on bin/generate-gutenboarding-design-thumbnails.mjs:
```diff
@@ -12,24 +12,19 @@
- * `node ./bin/generate-gutenboarding-design-thumbnails.mjs`
+ * `node ./bin/generate-gutenboarding-design-thumbnails.js`
  *
  * We should bump the version query param on the image src to cache bust the images too in client/landing/gutenboarding/available-designs.ts
```
Shouldn't we update also this comment to mention the new path updated in https://github.com/Automattic/wp-calypso/pull/51306/commits/66c4677e0c3c52374e3d1b51bfd316eb5954bb3d ? 

The change it's non-blocking and be added to https://github.com/Automattic/wp-calypso/issues/51404

ciampo replied:
Updated in https://github.com/Automattic/wp-calypso/pull/51306/commits/3e193ad66194537db7383f33c91352547199045f

razvanpapadopol reviewed (approved):
@ciampo, @StefanNieuwenhuis, awesome job extracting and packaging `@automattic/design-picker` 📦 

The `<DesignPicker />` component, which is the main goal of this PR, also looks really good for its first iteration outside Gutenboarding 🎉 

I've done more testing and all these work fine:
✅ `/new/design` si showing all the designs from the JSON config
✅ `/new/design?fresh` is showing designs in a different order (randomization works)
✅ visiting `/new/design` again is showing designs in the same order (persistence works)
✅ sites created with `/new` flow have the correct design and fonts applied
✅ the script that generates the design screenshots works as expected

ciampo committed:
Add jest-canvas-mock

ciampo committed:
Fix typo

ciampo committed:
Wrap available-designs-config.json into a ts import for the rest of the project

ciampo committed:
Add basic unit tests for utils

ciampo committed:
Delete placeholder tests

ciampo committed:
Update comment

ciampo commented:
> The <DesignPicker /> component, which is the main goal of this PR, also looks really good for its first iteration outside Gutenboarding 🎉

Thank you for reviewing @razvanpapadopol ! I've spent some time adding basic unit tests for the `utils` functions too

razvanpapadopol commented:
> basic unit tests for the utils functions too

Looking good!
<img width="1059" alt="Screenshot 2021-03-26 at 18 01 03" src="https://user-images.githubusercontent.com/14192054/112659702-4d9a6980-8e5d-11eb-8c17-e17db000f975.png">
